gson/src/main/java/com/google/gson/Gson.java

```diff
@@ -921,27 +921,19 @@ private static void assertFullConsumption(Object obj, JsonReader reader) {
    */
   @SuppressWarnings("unchecked")
   public <T> T fromJson(JsonReader reader, Type typeOfT) throws JsonIOException, JsonSyntaxException {
-    boolean isEmpty = true;
     boolean oldLenient = reader.isLenient();
     reader.setLenient(true);
     try {
-      JsonToken token = reader.peek();
-      if (token == JsonToken.END_DOCUMENT) {
-        return null; // we can also throw EOFException here but...
+      // For compatibility with JSON 1.5 and earlier, we return null for empty
+      // documents instead of throwing.
+      if (reader.peek() == JsonToken.END_DOCUMENT) {
+        return null;
       }
-      isEmpty = false;
       TypeToken<T> typeToken = (TypeToken<T>) TypeToken.get(typeOfT);
       TypeAdapter<T> typeAdapter = getAdapter(typeToken);
       T object = typeAdapter.read(reader);
       return object;
     } catch (EOFException e) {
-      /*
-       * For compatibility with JSON 1.5 and earlier, we return null for empty
-       * documents instead of throwing.
-       */
-      if (isEmpty) {
-        return null;
-      }
       throw new JsonSyntaxException(e);
     } catch (IllegalStateException e) {
```

Comment:
You can even remove this complete `catch` block. `EOFException` is an `IOException` and that is already caught below.

Reply:
Good point.